casper/src/main/scala/coop/rchain/casper/api/BlockApiImpl.scala

```diff
@@ -175,6 +183,81 @@ class BlockApiImpl[F[_]: Concurrent: RuntimeManager: BlockDagStorage: BlockStore
       )
   }
 
+  override def deployStatus(deployId: DeployId): F[ApiErr[DeployExecStatus]] = {
+    def notProcessed(status: String): DeployExecStatus =
+      DeployExecStatus().withNotProcessed(NotProcessed(status))
+
+    def findProcessedDeployResultAndStatus: OptionT[F, DeployExecStatus] = {
+      val lookupDeploy = OptionT(BlockDagStorage[F].lookupByDeployId(deployId))
+      lookupDeploy.semiflatMap { blockHash =>
+        val deployIdCh = DeployId(deployId.toByteArray)
+        for {
+          block     <- BlockStore[F].getUnsafe(blockHash)
+          deployOpt = block.body.deploys.find(_.deploy.sig == deployId)
+          deploy <- deployOpt.liftTo {
+                     val blockHashStr = PrettyPrinter.buildString(blockHash)
+                     val deploySigStr = PrettyPrinter.buildString(deployId)
+                     val errMsg =
+                       s"Deploy not found in the block, blockHash: $blockHashStr, deploy sig: $deploySigStr"
+                     new Exception(errMsg)
+                   }
+          status <- if (!deploy.isFailed)
+                     // For successfully executed deploy, get the result and block info
+                     for {
+                       data              <- getDataAtParRaw(deployIdCh, blockHash)
+                       (par, lightBlock) = data
+                       success           = ProcessedWithSuccess(par, lightBlock)
+                     } yield DeployExecStatus().withProcessedWithSuccess(success)
+                   else
+                     for {
+                       // For failed deploy, get the error and block info
+                       deployStatusOpt <- executionTracker.findDeploy(deployId)
+                       lightBlock      = getLightBlockInfo(block)
+                       errMsg = deployStatusOpt
+                         .collect { case DeployStatusError(errorMsg) => errorMsg }
+                         .getOrElse(
+                           "<deploy error message not available in cache or deploy executed on another node>"
+                         )
+                       error = ProcessedWithError(errMsg, lightBlock)
+                     } yield DeployExecStatus().withProcessedWithError(error)
+        } yield status
+      }
+    }
+
+    def findPooledDeploy: OptionT[F, DeployExecStatus] = {
+      val deployPooled = BlockDagStorage[F].containsDeployInPool(deployId).map(_.guard[Option])
+      OptionT(deployPooled).as {
+        // Deploy found in the pool, waiting to be executed and added to a block
+        notProcessed("Pooled")
+      }
+    }
+
+    def findCurrentlyExecutedDeploy: OptionT[F, DeployExecStatus] = {
+      val deployStatusOptT = OptionT(executionTracker.findDeploy(deployId))
+      deployStatusOptT.map { _ =>
+        // Status found, block creation in progress, deploy execution started
+        notProcessed("Running")
```

Comment:
Maybe we can use `as` function here as above in `findPooledDeploy`?
```scala
deployStatusOptT.as {
  // Status found, block creation in progress, deploy execution started
  notProcessed("Running")
}
```

Reply:
Yes, you are right. Good catch!
It can be on liner.
`deployStatusOptT.as(notProcessed("Running"))`